Source/lighting.h

```diff
@@ -74,10 +75,113 @@ void ChangeVisionXY(int id, Point position);
 void ProcessVisionList();
 void lighting_color_cycling();
 
+template <typename F>
+auto CrawlFlipsX(Displacement mirrored, F function) -> invoke_result_t<decltype(function), Displacement>
+{
+	const Displacement Flips[] = { mirrored.flipX(), mirrored };
+	for (auto displacement : Flips) {
+		auto ret = function(displacement);
+		if (ret)
+			return ret;
+	}
+	return {};
+}
+
+template <typename F>
+auto CrawlFlipsY(Displacement mirrored, F function) -> invoke_result_t<decltype(function), Displacement>
+{
+	const Displacement Flips[] = { mirrored, mirrored.flipY() };
+	for (auto displacement : Flips) {
+		auto ret = function(displacement);
+		if (ret)
+			return ret;
+	}
+	return {};
+}
+
+template <typename F>
+auto CrawlFlipsXY(Displacement mirrored, F function) -> invoke_result_t<decltype(function), Displacement>
+{
+	const Displacement Flips[] = { mirrored.flipX(), mirrored, mirrored.flipXY(), mirrored.flipY() };
+	for (auto displacement : Flips) {
+		auto ret = function(displacement);
+		if (ret)
+			return ret;
+	}
+	return {};
+}
+
+constexpr int MaxCrawlRadius = 18;
+
+/**
+ * CrawlTable specifies X- and Y-coordinate deltas from a missile target coordinate.
+ *
+ * n=4
+ *
+ *    y
+ *    ^
+ *    |  1
+ *    | 3#4
+ *    |  2
+ *    +-----> x
+ *
+ * n=16
+ *
+ *    y
+ *    ^
+ *    |  314
+ *    | B7 8C
+ *    | F # G
+ *    | D9 AE
+ *    |  526
+ *    +-------> x
+ */
+
+template <typename F>
+auto Crawl(unsigned radius, F function) -> invoke_result_t<decltype(function), Displacement>

```

Comment:
Is there any reason this is unsigned if we are just going to cast it to `int` anyway?

Reply:
it's cast to signed for use in a displacement that gets flipped and rotated, but the loops are expected to be always going from 0 (or a low positive value) to a larger value. Given the two contexts it's being used I think taking unsigned is appropriate here.